src/lib/storage/segment_access_statistics.hpp

```diff
@@ -0,0 +1,379 @@
+                  {State::RndInc, State::RndInc, State::RndInc, State::Rnd, State::Rnd},
+                  {State::SeqDec, State::Rnd, State::Rnd, State::SeqDec, State::RndDec},
+                  {State::RndDec, State::Rnd, State::Rnd, State::RndDec, State::RndDec},
+                  {State::Rnd, State::Rnd, State::Rnd, State::Rnd, State::Rnd}}};
+};
+// -----------------------------------------------------------------------------------------------------------------
+class AtomicAccessStrategy {
+ public:
+  uint64_t count(SegmentAccessType type) const;
+
+  void reset();
+
+  void increase(SegmentAccessType type, uint64_t count);
+
+  static std::string header();
+
+  std::vector<std::string> to_string() const;
+
+ private:
+  std::array<std::atomic_uint64_t, static_cast<uint32_t>(SegmentAccessType::Count)> _count;
+};
+
+// -----------------------------------------------------------------------------------------------------------------
+class NoAccessStrategy {
+ public:
+  uint64_t count(SegmentAccessType type) const;
+
+  void reset();
+
+  void increase(SegmentAccessType type, uint64_t count);
+
+  static std::string header();
+
+  std::vector<std::string> to_string() const;
+};
+
+// -----------------------------------------------------------------------------------------------------------------
+class AtomicTimedAccessStrategy {
+ public:
+  explicit AtomicTimedAccessStrategy();
+
+  uint64_t count(SegmentAccessType type) const;
+
+  void reset();
+
+  void increase(SegmentAccessType type, uint64_t count);
+
+  static std::string header();
+
+  std::vector<std::string> to_string() const;
+
+  static std::chrono::time_point<std::chrono::steady_clock> start_time;
+  static std::chrono::duration<double, std::milli> interval;
+
+ private:
+
+  struct Counter {
+    uint64_t timestamp;
+    std::array<std::atomic_uint64_t, static_cast<uint32_t>(SegmentAccessType::Count)> counter;
+
+    Counter(uint64_t timestamp) : timestamp{timestamp}, counter{} {};
+    Counter() : Counter(0ul) {}
+    Counter(const Counter& other) {
+      timestamp = other.timestamp;
+      for (auto type = 0ul; type < counter.size(); ++type) {
+        counter[type] = other.counter[type].load();
+      }
+    }
+    Counter(Counter&& other) noexcept {
+      timestamp = other.timestamp;
+      for (auto type = 0ul; type < counter.size(); ++type) {
+        counter[type] = other.counter[type].load();
+      }
+    }
+  };
+
+//  struct Counter {
+//    uint64_t timestamp;
+//    std::array<uint64_t, static_cast<uint32_t>(SegmentAccessType::Count)> counter;
+//
+//    Counter(uint64_t timestamp) : timestamp{timestamp}, counter{} {};
+//    Counter() : Counter(0ul) {}
+//  };
+
+  pmr_concurrent_vector<Counter> _counters;
+  Counter& _counter();
+};
+
+// -----------------------------------------------------------------------------------------------------------------
+class NonLockingStrategy {
```

Comment:
Only one of these will survive at the end, correct?

Reply:
All will be gone. I have the feeling that accessing the counters over the strategies has a visible performance impact.